dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp

```diff
@@ -389,9 +441,293 @@ SegmentPtr DeltaMergeStore::segmentMerge(
+
+    LOG_INFO(log, "CheckAllSegmentsLocalIndex - Begin");
+
+    size_t segments_updated_meta = 0;
+    auto dm_context = newDMContext(global_context, global_context.getSettingsRef(), "checkAllSegmentsLocalIndex");
+
+    // 1. Make all segments referencing latest meta version.
+    {
+        Stopwatch watch;
+        std::unique_lock lock(read_write_mutex);
+
+        std::map<PageIdU64, DMFilePtr> latest_dmf_by_id;
+        for (const auto & [end, segment] : segments)
+        {
+            UNUSED(end);
+            for (const auto & dm_file : segment->getStable()->getDMFiles())
+            {
+                auto & latest_dmf = latest_dmf_by_id[dm_file->fileId()];
+                if (!latest_dmf || dm_file->metaVersion() > latest_dmf->metaVersion())
+                    // Note: pageId could be different. It is fine.
+                    latest_dmf = dm_file;
+            }
+        }
+        for (const auto & [end, segment] : segments)
+        {
+            UNUSED(end);
+            for (const auto & dm_file : segment->getStable()->getDMFiles())
+            {
+                auto & latest_dmf = latest_dmf_by_id.at(dm_file->fileId());
+                if (dm_file->metaVersion() < latest_dmf->metaVersion())
+                {
+                    // Note: pageId could be different. It is fine, replaceStableMetaVersion will fix it.
+                    auto update_result = segmentUpdateMeta(lock, *dm_context, segment, {latest_dmf});
+                    RUNTIME_CHECK(update_result != nullptr, segment->simpleInfo());
+                    ++segments_updated_meta;
+                }
+            }
+        }
+        LOG_INFO(
+            log,
+            "CheckAllSegmentsLocalIndex - Finish, updated_meta={}, elapsed={}",
+            segments_updated_meta,
+            watch.elapsedSeconds());
+    }
+
+    size_t segments_missing_indexes = 0;
+
+    // 2. Trigger ensureStableIndex for all segments.
+    // There could be new segments between 1 and 2, which is fine. New segments
+    // will invoke ensureStableIndex at creation time.
+    {
+        // There must be a lock, because segments[] may be mutated.
+        // And one lock for all is fine, because segmentEnsureStableIndexAsync is non-blocking, it
+        // simply put tasks in the background.
+        std::shared_lock lock(read_write_mutex);
+        for (const auto & [end, segment] : segments)
+        {
+            UNUSED(end);
+            if (segmentEnsureStableIndexAsync(segment))
+                ++segments_missing_indexes;
+        }
+    }
+
+    LOG_INFO(
+        log,
+        "CheckAllSegmentsLocalIndex - Finish, segments_[updated_meta/missing_index]={}/{}",
+        segments_updated_meta,
+        segments_missing_indexes);
+}
+
+bool DeltaMergeStore::segmentEnsureStableIndexAsync(const SegmentPtr & segment)
+{
+    RUNTIME_CHECK(segment != nullptr);
+
+    // TODO(local index): There could be some indexes are built while some indexes is not yet built after DDL
```

Comment:
There could be some indexes are **being** built?

Reply:
Just ignore it, will remove it in a later PR.